compiler/src/iree/compiler/Preprocessing/Common/ConvertConvFilterToChannelsLast.cpp

```diff
@@ -0,0 +1,144 @@
+  Value filter = convOp.getInputs()[1];
+  Value output = convOp.getOutputs()[0];
+
+  // Extract indexing maps.
+  AffineMap inputMap = convOp.getIndexingMapsArray()[0];
+  AffineMap filterMap = convOp.getIndexingMapsArray()[1];
+  AffineMap outputMap = convOp.getIndexingMapsArray()[2];
+
+  AffineMap transposedFilterMap = constructFilterMap(filterMap, perm);
+  Value transposedFilter = createTransposeOp(rewriter, loc, filter, perm);
+
+  SmallVector<utils::IteratorType> iterators = convOp.getIteratorTypesArray();
+
+  auto genericOp = rewriter.create<linalg::GenericOp>(
+      loc, output.getType(), ValueRange{input, transposedFilter}, output,
+      ArrayRef<AffineMap>{inputMap, transposedFilterMap, outputMap}, iterators);
+
+  // Reuse the same payload as the original convolution op.
+  rewriter.inlineRegionBefore(convOp->getRegion(0), genericOp.getRegion(),
+                              genericOp.getRegion().begin());
+
+  rewriter.replaceOp(convOp, genericOp->getResults());
+  return success();
+}
+
+namespace {
+struct ConvertHwcfToHwfc : public OpRewritePattern<linalg::Conv2DNhwcHwcfOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<int64_t> perm = {0, 1, 3, 2};
+    return convertConvFilterToTargetLayout(convOp, rewriter, perm);
+  }
+};
+
+struct ConvertHwcfToFhwc : public OpRewritePattern<linalg::Conv2DNhwcHwcfOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<int64_t> perm = {3, 0, 1, 2};
+    return convertConvFilterToTargetLayout(convOp, rewriter, perm);
+  }
+};
+
+class ConvertConvFilterToChannelsLastPass
+    : public iree_compiler::Preprocessing::impl::
+          ConvertConvFilterToChannelsLastPassBase<
+              ConvertConvFilterToChannelsLastPass> {
+public:
+  using iree_compiler::Preprocessing::impl::
+      ConvertConvFilterToChannelsLastPassBase<
+          ConvertConvFilterToChannelsLastPass>::
+          ConvertConvFilterToChannelsLastPassBase;
+
+  void runOnOperation() override {
+    auto op = getOperation();
+    MLIRContext *context = &getContext();
+
+    RewritePatternSet patterns(context);
+    if (filterLayout == "hwfc") {
+      patterns.add<ConvertHwcfToHwfc>(context);
+    } else if (filterLayout == "fhwc") {
+      patterns.add<ConvertHwcfToFhwc>(context);
+    } else {
+      // TODO add default fallback to filter layout once we have more data
+      // about models with the two layouts
+      llvm_unreachable("Unsupported filter layout. Use hwfc or fhwc.");
+    }
+
+    if (failed(applyPatternsGreedily(op, std::move(patterns)))) {
+      return signalPassFailure();
+    }
+
+    LDBG("after converting convolutions to channels last\n" << *op);
```

Comment:
IMO, this is an unnecessary log. Because this is the result after running the pass and users can just use `-mlir-print-ir-after=...` to get the information.